### PR TITLE
Add a removeAll pathConf cmd to support remove all path conf

### PR DESCRIPTION
**Pull request by maobaolong**
### What changes are proposed in this pull request?

Please outline the changes and how this PR fixes the issue.

### Why are the changes needed?

Please clarify why the changes are needed. For instance,
  1. If you propose a new API, clarify the use case for a new API.
  2. If you fix a bug, describe the bug.

### Does this PR introduce any user facing changes?

Please list the user-facing changes introduced by your change, including
  1. change in user-facing APIs
  2. addition or removal of property keys
  3. webui

### CODE REVIEW

**Review by elega — APPROVED** [review]

**Comment by dbw9580** [comment]
I'd propose a different approach to do this. Since there is already a `remove` command, how about we add a `--recursive` option to `remove`, which will recursively removes all path configurations starting with the prefix specified by the positional argument.

For example:

```console
bin/alluxio fsadmin pathConf remove -r /
```

will recursively remove all path confs set on any level, since it's the root path

```console
bin/alluxio fsadmin pathConf remove -r --key "key1,key2" /
```

will recursively remove path confs with the keys "key1" and "key2" on any level

Contrarily, without the `-r` option, only the configuration set on the specific level will be removed:

```console
bin/alluxio fsadmin pathConf remove --key "key1,key2" /
```

will only remove key "key1" and "key2" on the root path, but leaving the same configurations set on any child paths intact. (which **should** be the current behavior, if not, please leave a comment.)

Finally, the positional argument should be processed on a path component granularity, rather than character granularity, i.e. `/tmp` is *not* a prefix of `/tmp1/dir`, etc.

File: shell/src/main/java/alluxio/cli/fsadmin/pathconf/RemoveCommand.java [hunk 1]
```diff
@@ -42,6 +45,11 @@ public final class RemoveCommand extends AbstractFsAdminCommand {
           .numberOfArgs(1)
           .desc("properties keys to be removed from this path's configurations, separated by comma")
           .build();
+  private static final Option RECURSIVE_OPTION =
+      Option.builder("R")
+          .longOpt("recursive")
+          .desc("Remove all properties starts with the given key prefix")
```
Comment (dbw9580) [inline shell/src/main/java/alluxio/cli/fsadmin/pathconf/RemoveCommand.java:51]:
```suggestion
          .desc("remove properties set on paths starting from the specified path recursively")
```

File: shell/src/main/java/alluxio/cli/fsadmin/pathconf/RemoveCommand.java [hunk 2]
```diff
@@ -42,6 +45,11 @@ public final class RemoveCommand extends AbstractFsAdminCommand {
           .numberOfArgs(1)
           .desc("properties keys to be removed from this path's configurations, separated by comma")
```
Comment (dbw9580) [inline shell/src/main/java/alluxio/cli/fsadmin/pathconf/RemoveCommand.java:46]:
```suggestion
          .desc("property keys to be removed, separated by comma")
```

File: shell/src/main/java/alluxio/cli/fsadmin/pathconf/RemoveCommand.java [hunk 3]
```diff
@@ -69,19 +77,37 @@ public void validateArgs(CommandLine cl) throws InvalidArgumentException {
...
+    boolean recursive = cl.hasOption(RECURSIVE_OPTION.getOpt());
     if (cl.hasOption(KEYS_OPTION_NAME)) {
       String[] keys = cl.getOptionValue(KEYS_OPTION_NAME).split(",");
       Set<PropertyKey> propertyKeys = new HashSet<>();
       for (String key : keys) {
         propertyKeys.add(PropertyKey.fromString(key));
       }
-      mMetaConfigClient.removePathConfiguration(path, propertyKeys);
+      handleRemove(path, recursive,
+          p -> mMetaConfigClient.removePathConfiguration(p, propertyKeys));
```
Comment (dbw9580) [inline shell/src/main/java/alluxio/cli/fsadmin/pathconf/RemoveCommand.java:88]: this is a bit too convoluted. I think it's best to structure it as:

```java
Set<PropertyKey> keysToRemove = ...;
Set<AlluxioURI> pathsToRemove = isRecursive ? ... : emptySet();
if (keysToRemove.isEmpty()) {
  // remove all keys
  for (AlluxioURI p : pathsToRemove) {
    client.removePathConfiguration(p);
  }
  return 0;
}

for (AlluxioURI p : pathsToRemove) {
  client.removePathConfiguration(p, keysToRemove);
}

return 0;
```

File: shell/src/main/java/alluxio/cli/fsadmin/pathconf/RemoveCommand.java [hunk 4]
```diff
@@ -69,19 +77,37 @@ public void validateArgs(CommandLine cl) throws InvalidArgumentException {
...
+      handleRemove(path, recursive, p -> mMetaConfigClient.removePathConfiguration(p));
     }
     return 0;
   }
 
+  private void handleRemove(AlluxioURI path, boolean recursive,
+      CheckedConsumer<AlluxioURI> function) throws IOException {
+    if (recursive) {
+      Configuration conf = mMetaConfigClient.getConfiguration(
+          GetConfigurationPOptions.newBuilder().setIgnoreClusterConf(true).build());
```
Comment (dbw9580) [inline shell/src/main/java/alluxio/cli/fsadmin/pathconf/RemoveCommand.java:99]: why `setIgnoreClusterConf(true)`?

File: shell/src/main/java/alluxio/cli/fsadmin/pathconf/RemoveCommand.java [hunk 5]
```diff
@@ -69,19 +77,37 @@ public void validateArgs(CommandLine cl) throws InvalidArgumentException {
...
+        }
+      }
+    } else {
+      function.consume(path);
+    }
+  }
+
   @Override
   public String getUsage() {
     return String.format("%s [--%s <key1,key2,key3>] <path>%n"
```
Comment (dbw9580) [inline shell/src/main/java/alluxio/cli/fsadmin/pathconf/RemoveCommand.java:113]: please update the usage string as well.

**Review by dbw9580 — CHANGES_REQUESTED** [review]

File: shell/src/main/java/alluxio/cli/fsadmin/pathconf/RemoveCommand.java (hunk 4 above)
Comment (maobaolong) [reply shell/src/main/java/alluxio/cli/fsadmin/pathconf/RemoveCommand.java:99]: Just copy it from `ListCommand.java`

File: shell/src/main/java/alluxio/cli/fsadmin/pathconf/RemoveCommand.java (hunk 5 above)
Comment (maobaolong) [reply shell/src/main/java/alluxio/cli/fsadmin/pathconf/RemoveCommand.java:113]: done

File: shell/src/main/java/alluxio/cli/fsadmin/pathconf/RemoveCommand.java (hunk 3 above)
Comment (maobaolong) [reply shell/src/main/java/alluxio/cli/fsadmin/pathconf/RemoveCommand.java]: okay, i turn to refactor the code follow your style

**Comment by maobaolong** [comment]
@dbw9580 Thanks for your review and suggestion. PTAL

**Comment by maobaolong** [comment]
@dbw9580 Thanks for the review comments, I changed the code style, PTAL.

**Comment by dbw9580** [comment]
@maobaolong the implementation looks good to me. Can you please add tests for this new feature in `alluxio.client.cli.fsadmin.pathconf.RemoveCommandIntegrationTest`?

**Comment by maobaolong** [comment]
@dbw9580 Thank you for your information, I add a case test to the current UT. PTAL

File: tests/src/test/java/alluxio/client/cli/fsadmin/pathconf/RemoveCommandIntegrationTest.java [hunk 6]
```diff
@@ -64,6 +64,25 @@ public void remove() throws Exception {
       output = mOutput.toString();
       Assert.assertEquals(DIR1 + "\n" + DIR2 + "\n", output);
 
+      ret = shell.run("pathConf", "remove", "-R", "/");
```
[the remark this answers is not in the document]
Comment (maobaolong) [reply tests/src/test/java/alluxio/client/cli/fsadmin/pathconf/RemoveCommandIntegrationTest.java]: done

**Comment by maobaolong** [comment]
@dbw9580 PTAL.

File: shell/src/main/java/alluxio/cli/fsadmin/pathconf/RemoveCommand.java [hunk 7]
```diff
@@ -69,22 +78,39 @@ public void validateArgs(CommandLine cl) throws InvalidArgumentException {
...
     return 0;
   }
 
   @Override
   public String getUsage() {
-    return String.format("%s [--%s <key1,key2,key3>] <path>%n"
+    return String.format("%s [-R/--recursive] [--%s <key1,key2,key3>] <path>%n"
         + "\t--%s: %s",
         getCommandName(), KEYS_OPTION_NAME,
         KEYS_OPTION_NAME, KEYS_OPTION.getDescription());
```
Comment (dbw9580) [inline shell/src/main/java/alluxio/cli/fsadmin/pathconf/RemoveCommand.java:116]: also add the recursive option to the usage

**Review by dbw9580 — APPROVED** [review]
LGTM, thanks for adding this feature!

File: core/client/fs/src/main/java/alluxio/client/file/ConfigHashSync.java [hunk 8]
```diff
@@ -86,7 +86,7 @@ public synchronized void heartbeat() {
       return;
     }
     boolean isClusterConfUpdated = !hash.getClusterConfigHash().equals(
-        mContext.getClusterConf().hash());
+        mContext.getClientContext().getClusterConfHash());
```
[the remark this answers is not in the document]
Comment (maobaolong) [reply core/client/fs/src/main/java/alluxio/client/file/ConfigHashSync.java:89]: Yeah...

File: underfs/hdfs/src/main/java/alluxio/underfs/hdfs/HdfsUnderFileSystem.java [hunk 9]
```diff
@@ -450,21 +450,21 @@ public long getSpace(String path, SpaceType type) throws IOException {
       switch (type) {
         case SPACE_TOTAL:
           //#ifdef HADOOP1
-          space = ((DistributedFileSystem) hdfs).getDiskStatus().getCapacity();
+//          space = ((DistributedFileSystem) hdfs).getDiskStatus().getCapacity();
```
[the remark this answers is not in the document]
Comment (maobaolong) [reply underfs/hdfs/src/main/java/alluxio/underfs/hdfs/HdfsUnderFileSystem.java]: Done

File: shell/src/main/java/alluxio/cli/fsadmin/pathconf/RemoveCommand.java (hunk 7 above)
Comment (maobaolong) [reply shell/src/main/java/alluxio/cli/fsadmin/pathconf/RemoveCommand.java]: Good suggestion!

**Comment by maobaolong** [comment]
@dbw9580 Thanks for the suggestion, I reset the previous commit and push a new one to revert the Confighash related code which is not related to this PR, also, i added the recursive usage to the usage method. PTAL.

File: shell/src/main/java/alluxio/cli/fsadmin/pathconf/RemoveCommand.java [hunk 10]
```diff
@@ -69,25 +78,43 @@ public void validateArgs(CommandLine cl) throws InvalidArgumentException {
...
   @Override
   public String getUsage() {
-    return String.format("%s [--%s <key1,key2,key3>] <path>%n"
-        + "\t--%s: %s",
+    return String.format("%s [-R/--recursive] [--%s <key1,key2,key3>] <path>%n"
+        + "\t--%s: %s" + "\t--%s: %s",
         getCommandName(), KEYS_OPTION_NAME,
-        KEYS_OPTION_NAME, KEYS_OPTION.getDescription());
+        KEYS_OPTION_NAME, KEYS_OPTION.getDescription(),
+        RECURSIVE_OPTION, RECURSIVE_OPTION.getDescription());
```
Comment (dbw9580) [inline shell/src/main/java/alluxio/cli/fsadmin/pathconf/RemoveCommand.java:117]: `RECURSIVE_OPTION` is an `Option` not a `String`.

```suggestion
        RECURSIVE_OPTION.getOpt(), RECURSIVE_OPTION.getDescription());
```

File: shell/src/main/java/alluxio/cli/fsadmin/pathconf/RemoveCommand.java [hunk 11]
```diff
@@ -69,25 +78,43 @@ public void validateArgs(CommandLine cl) throws InvalidArgumentException {
...
     }
     return 0;
   }
 
   @Override
   public String getUsage() {
-    return String.format("%s [--%s <key1,key2,key3>] <path>%n"
-        + "\t--%s: %s",
+    return String.format("%s [-R/--recursive] [--%s <key1,key2,key3>] <path>%n"
+        + "\t--%s: %s" + "\t--%s: %s",
```
Comment (dbw9580) [inline shell/src/main/java/alluxio/cli/fsadmin/pathconf/RemoveCommand.java:114]:
```suggestion
        + "\t--%s: %s" + "%n"
        + "\t--%s: %s",
```

**Review by dbw9580 — APPROVED** [review]

**Comment by maobaolong** [comment]
@elega @dbw9580 Thanks for the review, as all reviewers approved this PR, I merge this PR now.

**Comment by maobaolong** [comment]
alluxio-bot, merge this please

**Pull request merged**